ullyses/ullyses_hlsp.py

```diff
@@ -599,37 +770,82 @@ def combine_keys(self, key, method, dict_key=None, constant=None):
+                         "opt_elem": ("seqarm", 1),
+                         "s_region": ("hierarch eso ada posang", 1), # calculate this from position angle
+                         "equinox": ("equinox", 1),
+                         "radesys": ("radecsys", 1),
+                         "proposid": ("hierarch eso obs prog id", 1),
+                         "centrwv": ("arm", 1), 
+                         "minwave": ("arm", 1), 
+                         "maxwave": ("arm", 1), 
+                         "filename": ("extname", 1),
+                         "dr_num": ("dr_num", 0),
+                         "dr_date": ("dr_date", 0),
+                         "specres": ("special", 1),
+                         "cal_ver": ("hierarch eso pro rec1 pipe id", 1)}
                          }
 
         if constant is not None:
-            vals = [constant for x in self.primary_headers]
+            vals = [constant for x in self.first_headers]
         else:
             vals = []
-            for i in range(len(self.primary_headers)):
+            for i in range(len(self.first_headers)):
                 if dict_key is None:
-                    tel = self.primary_headers[i]["telescop"]
+                    tel = self.telescope
                 else:
                     tel = dict_key
-                actual_key = keymap[tel][key][0]
-                hdrno = keymap[tel][key][1]
-                if hdrno == 0:
-                    val = self.primary_headers[i][actual_key]
-                else:
-                    val = self.first_headers[i][actual_key]
-                if tel == "FUSE" and key == "filename":
-                    val = val.replace(".fit", "_vo.fits")
-                if tel == "LCOGT":
-                    dto = dt.strptime(self.first_headers[i]["date-obs"], "%Y-%m-%dT%H:%M:%S.%f")
-                    t = Time(dto, format="datetime")
-                    mjdstart = t.mjd
-                    if key == "expstart":
-                        val = mjdstart
-                    if key == "expend":
+                
+                match [tel, key]:
+                    # Handle some special cases
+                    case ["FUSE", "filename"]:
+                        val = val.replace(".fit", "_vo.fits")
+                    case ["LCOGT", "telescop"]:
+                        telescop = self.first_headers[i]["telescop"] 
+                        val = f"LCOGT-{telescop}"
+                    case ["LCOGT", "expstart" | "expend" as k]:
+                        dto = dt.strptime(self.first_headers[i]["date-obs"], "%Y-%m-%dT%H:%M:%S.%f")
+                        t = Time(dto, format="datetime")
+                        mjdstart = t.mjd
+                        if k == "expstart":
+                            val = mjdstart
+                        if k == "expend":
+                            exptime = self.first_headers[i]["exptime"]
+                            val = mjdstart + (exptime/86400) # seconds in a day
```

Comment:
It's not a bad idea to make a global variable at the top:

SECONDS_PER_DAY = 86400.

and then you can use it in the code body instead of 86400, then you don't need the inline comment